### PR TITLE
Feat:Staff : Create a new API for basic search

### PR DESCRIPTION
**Pull request by shivajihmcts**
### JIRA link  ###
https://tools.hmcts.net/jira/browse/RDCC-5443

### Change description ###
Staff : Create a new API for basic search

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        String searchString = "cwr";
+        when(caseWorkerProfileRepo.findByFirstNameOrLastName(searchString.toLowerCase(), pageRequest))
+                .thenReturn(pages);
+        ResponseEntity<List<SearchStaffUserResponse>> responseEntity =
+                staffRefDataServiceImpl.retrieveStaffUserByName(searchString, pageRequest);
+        assertEquals(200, responseEntity.getStatusCodeValue());
+
+        List<SearchStaffUserResponse> searchResponse =
+                responseEntity.getBody();
+        assertThat(responseEntity.getBody()).isNotNull();
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:164]: 1. can you add assertions for expected content in the response 
2. add assertion for total records in response header

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        validateSearchStaffUserResponses(searchResponse);
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:135]: check  response header has total records and its value.

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                .andReturn();
+        fetchResponse.then()
+                .assertThat()
+                .statusCode(200);
+
+        List<SearchStaffUserResponse> searchStaffUserResponse = Arrays.asList(
+                fetchResponse.getBody().as(SearchStaffUserResponse[].class));
+        assertThat(searchStaffUserResponse).isNotNull();
+        assertThat(searchStaffUserResponse.get(0).getFirstName()).contains(searchString);
+        assertThat(searchStaffUserResponse.get(0).getLastName()).contains(searchString);
```
Comment (rajkatla-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java:109]: check response header has total records

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/repository/CaseWorkerProfileRepository.java [hunk 4]
```diff
@@ -24,6 +24,11 @@ public interface CaseWorkerProfileRepository extends JpaRepository<CaseWorkerPro
 
     List<CaseWorkerProfile> findByEmailIdIgnoreCaseContaining(String emailPattern);
 
+    @Query(value = "select cw from case_worker_profile cw where lower(cw.firstName) like "
+            + "concat('%', :searchString, '%') "
```
Comment (rajkatla-hmcts) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/repository/CaseWorkerProfileRepository.java:28]: can this query support full first name or last name search ? and how the records will be order by last name , first name asc ?

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+        String searchString = "cwr";
+
+
+        Response response = caseWorkerApiClient.withUnauthenticatedRequest()
+                .get(CASE_WORKER_PROFILE_URL + "/search-by-name?search=" + searchString)
+                .andReturn();
+        response.then()
+                .assertThat()
+                .statusCode(401);
```
[the remark this answers is not in the document]
Comment (shivajihmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java:144]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.reform.cwrdapi;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import uk.gov.hmcts.reform.cwrdapi.util.AuthorizationEnabledIntegrationTest;
+
+class StaffRefIntegrationTest extends AuthorizationEnabledIntegrationTest {
```
[the remark this answers is not in the document]
Comment (shivajihmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefIntegrationTest.java:8]: done

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 7]
```diff
@@ -41,9 +55,65 @@ public class StaffRefDataController {
...
+        ),
+        @ApiResponse(
+            code = 500,
+            message = INTERNAL_SERVER_ERROR
+        )
+    })
+    @Validated
+    @GetMapping(path = "/profile/search-by-name",
+                produces = APPLICATION_JSON_VALUE)
+    @Secured("cwd-admin")
```
[the remark this answers is not in the document]
Comment (shivajihmcts) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java]: done

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/repository/CaseWorkerProfileRepository.java (hunk 4 above)
Comment (shivajihmcts) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/repository/CaseWorkerProfileRepository.java:28]: Its done in page at controller

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java (hunk 3 above)
Comment (shivajihmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java:109]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        String role = "cwd-admin";
+        String searchString = "";
+
+        Map<String, Object> response =  caseworkerReferenceDataClient
+                .searchStaffUserByName(path,searchString,0,20,role);
+
+        assertThat(response).containsEntry("http_status", "400");
+        assertThat(response.get("response_body").toString())
+                .contains("Required request parameter 'search' for method parameter type String is not present");
+
```
[the remark this answers is not in the document]
Comment (shivajihmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataIntegrationTest.java:146]: We are using integration test cases for client side errors validation and functional test cases for result validation.
Sorting and record counts are covered in functional test cases.

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    public static final String STAFF_REF_DATA_SKILL_URL = "/refdata/case-worker/skill";
+    public static final String STAFF_REF_DATA_RD_STAFF_UI_KEY =
+            "StaffRefDataController.retrieveAllServiceSkills";
+    public static final String SEARCH_STAFF_USER_BY_NAME_KEY = "StaffRefDataController.searchStaffUserByName";
+    public static final String CASE_WORKER_PROFILE_URL = "/refdata/case-worker/profile";
+
+    @Test
+    @ToggleEnable(mapKey = STAFF_REF_DATA_RD_STAFF_UI_KEY, withFeature = true)
+    @ExtendWith(FeatureToggleConditionExtension.class)
+    public void should_return_service_skills_with_status_code_200_when_flag_true() {
```
Comment (SujeethKumarAri) [inline src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java:49]: These skill related tests are already covered in https://github.com/hmcts/rd-caseworker-ref-api/pull/584/files#diff-8d51055287307302fe906e24c0670cc56d9851e86439519b44fb30eb05703f83R40 . May I know why you have included it again ?

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java (hunk 9 above)
Comment (shivajihmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataFunctionalTest.java]: removed skills related duplicate test cases

**Review by SujeethKumarAri — APPROVED** [review]

**Review by sahitya-desireddy — APPROVED** [review]

**Review by nayeemshaik-hmcts — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java (hunk 2 above)
Comment (shivajihmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:139]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java [hunk 10]
```diff
@@ -184,6 +185,77 @@ public Map<String, Object> fetchStaffProfileByCcdServiceName(String ccdServiceNa
...
+        stringBuilder
+                .append(path);
+
+        if (StringUtils.isNotBlank(searchString)) {
+            stringBuilder.append("?search=");
+            stringBuilder.append(searchString);
+        }
+
+        HttpHeaders headers =  getMultipleAuthHeadersWithPagination(role, null,pageNumber,pageSize);
+        //headers.add("page-number",pageNumber);
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:201]: pls remove commented lines

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/CreateStaffReferenceProfileBasicSearchTest.java [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    void createCaseWorkerTestData(String firstName, String lastName, String email) {
+        List<CaseWorkerRoleRequest> roleRequests = new ArrayList<CaseWorkerRoleRequest>();
+        roleRequests.add(new CaseWorkerRoleRequest("National Business Centre Team Leader", true));
+        roleRequests.add(new CaseWorkerRoleRequest("Regional Centre Team Leader", false));
+        List<CaseWorkersProfileCreationRequest> caseWorkersProfileCreationRequests =
+                createCaseWorkerProfiles(firstName, lastName, email);
+        caseWorkersProfileCreationRequests.get(0).setRoles(roleRequests);
+        //Response response = caseWorkerApiClient.createUserProfiles(caseWorkersProfileCreationRequests);
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/CreateStaffReferenceProfileBasicSearchTest.java:274]: pls remove commented line

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/CreateStaffReferenceProfileBasicSearchTest.java [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                .caseWorkerWorkAreaRequest()
+                .serviceCode("ABA1").areaOfWork("Divorce").build();
+
+        List<CaseWorkerWorkAreaRequest> areaRequests =
+                ImmutableList.of(workerWorkAreaRequest1, workerWorkAreaRequest2);
+
+        Set<String> idamRoles = new HashSet<>();
+
+        String emailToUsed = nonNull(email) ? email : generateRandomEmail();
+        // setEmailsTobeDeleted(emailToUsed.toLowerCase());
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/CreateStaffReferenceProfileBasicSearchTest.java:308]: pls remove commented line

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java (hunk 10 above)
Comment (shivajihmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java]: done

**Review by rajkatla-hmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/domain/CaseWorkerSkill.java [hunk 13]
```diff
@@ -54,6 +76,9 @@ public class CaseWorkerSkill implements Serializable {
             insertable = false, updatable = false, nullable = false)
     private CaseWorkerProfile caseWorkerProfile;
 
+    @Fetch(FetchMode.SUBSELECT)
+    @OneToMany(targetEntity = Skill.class, mappedBy = "caseWorkerSkill", cascade = ALL, orphanRemoval = true)
```
Comment (sahitya-desireddy) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/domain/CaseWorkerSkill.java:80]: Hi @shivajihmcts on further discussions with the Team, the above skill relationship in caseworker skills should be many to one Request you to please do the needful Thank you so much ... FYI @SujeethKumarAri @kavithagade12

Comment (shivajihmcts) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/domain/CaseWorkerSkill.java]: Hi Sahitya
Please check code on the staff-ui-changes, its have same mapping as you expecting, Please take a update from branch. Its a part of create staff profile , mapping is covered into it

**Pull request merged**